### DISCUSSION TITLE
UnloadModel exception when freeing VBO

### DISCUSSION THREAD

**Discussion by ShaharKlingerSSYS** [raysan5/raylib#2993] [Q&A] [answered]
Hello everyone,

I've been struggling with this for the last 2 days, and can't seem to find the cause: when calling `UnloadModel`, I get an exception thrown at line [1763](https://github.com/raysan5/raylib/blob/master/src/rmodels.c#L1763) in `rmodels.c` - that is the command `RL_FREE(mesh.vboId);`.

What can be the scenario or cause this line throws an exception? The model has only one mesh, the vertex array is unloaded successfully, `vboId` is not NULL.

Thanks for helping.

**Comment by raysan5** [comment]
@ShaharKlingerSSYS How are you loading/creating the model? Please, could you show the loading/creation code?

The issue could be related to `mesh.vboId` array not being properly initialized or maybe freed manually before `UnloadModel()`.

**Reply by ShaharKlingerSSYS** [reply]
Hey, thanks for answering quickly. I'm creating a model from a "flat" vector of vertices in this manner:

```
Mesh mesh = { 0 };
size_t vertexElementCount = vertices.size();
size_t vertexCount = vertexElementCount / 3;
size_t faceCount = vertexCount / 3;

// Vertices
size_t memorySize = vertexElementCount * sizeof(float);
mesh.vertices = (float*)malloc(memorySize);
memcpy(mesh.vertices, vertices.data(), memorySize);
mesh.vertexCount = static_cast<int>(vertexCount);

// Normals
mesh.normals = (float*)malloc(memorySize);
CalculateFlatShadingNormals(mesh.vertices, faceCount, mesh.normals);    // My function.

// Faces
mesh.indices = nullptr;
mesh.triangleCount = static_cast<int>(faceCount);

// Model
UploadMesh(&mesh, false);
Model model = LoadModelFromMesh(mesh);
```

The resulting model is saved as a data member of a "scene object" class.

Was there a change in this area of the raylib code between 4.5.0-dev and 4.5.0? Before upgrading I didn't have this issue. Then again, I built 4.5.0-dev myself, but took 4.5.0 from the GitHub release, so I don't know if it's related...

Thanks again.

**Reply by raysan5** [reply]
Afair, there was no change in those functions... Please, could you check values of `mesh.vboId[0] .. mesh.vboId[6]` after `UploadMesh()`?

**Reply by raysan5** [reply]
Just checking `UploadMesh()` and your code more carefully and I think it could be related to `UploadMesh()` expecting `mesh->texcoords` defined. You are not defining them.

**Reply by ShaharKlingerSSYS** [reply]
Well, strangely enough, this problem disappears in _release_ build, but persists in debug (usually it's the other way around...).

The values of `mesh.vboId` after uploading are: 74, 75, 76, 0, 0, 0, 0.

I tried adding all-zero texture coordinates to the mesh but that didn't solve the issue.

**Reply by raysan5** [reply]
That's really weird, it should be debugged carefully to see where the issue could be...

**Reply by ShaharKlingerSSYS** [reply]
Indeed, something is really weird. When I use the function `GenMeshPoly` there's no problem, but when I copy that function's code without change to my main program, it crashes when unloading the model.

I can confirm it happens with a small stand alone program, outside the project I'm working on. Maybe something to do with my project's setup...?

**Reply by raysan5** [reply, marked as answer]
Hey! I think I know the issue! You are using different allocators to load and free the memory! Try using raylib `MemAlloc()` for your vertex arrays instead of system provided `malloc()`.

**Reply by ShaharKlingerSSYS** [reply]
Wow, good call! That was indeed the problem. It's working perfectly now.

Thank you very much for your help! All the best.

**Reply by raysan5** [reply]
Nice! 👍😄
